File: 0556. Next Greater Element III/Solution.py
class Solution:
    def nextGreaterElement(self, n: int) -> int:
        s = [int(c) for c in str(n)]
        i = j = len(s) - 1
        while i > 0 and s[i] <= s[i - 1]:
            i -= 1
        if i == 0:
            return -1
        while s[j] <= s[i - 1]:
            j -= 1
        s[i - 1], s[j] = s[j], s[i - 1]
        s[i:] = reversed(s[i:])
        INT_MAX = 2 ** 31 - 1
        res = 0
        for c in s:
            if res > (INT_MAX - c) // 10:
                return -1
            res = res * 10 + c
        return res

File: 0556. Next Greater Element III/Solution.cpp
#include <algorithm>
#include <string>
using namespace std;

class Solution {
public:
    int nextGreaterElement(int n)
    {
        string s = to_string(n);
        if (!next_permutation(s.begin(), s.end()))
            return -1;
        int res = 0;
        for (const char& c : s) {
            int d = c - '0';
            if (res > (INT_MAX - d) / 10)
                return -1;
            res = res * 10 + d;
        }
        return res;
    }
};
